File: Ejercicio5_Transformacion.cpp
#include <iostream>
#include <algorithm> // std::max
using namespace std;

struct Nodo {
    int valor;
    Nodo* izquierdo;
    Nodo* derecho;
    Nodo(int v) : valor(v), izquierdo(nullptr), derecho(nullptr) {}
};

// Calcula la altura de un árbol binario (número de niveles)
// La altura es el camino más largo desde la raíz hasta una hoja
int calcularAltura(Nodo* raiz) {
    if (raiz == nullptr)  // Caso base: árbol vacío tiene altura 0
        return 0;

    int alturaIzq = calcularAltura(raiz->izquierdo);  // Altura del subárbol izquierdo
    int alturaDer = calcularAltura(raiz->derecho);    // Altura del subárbol derecho

    // La altura del nodo actual es 1 (el nodo mismo) +
    // el mayor entre ambos subárboles
    return 1 + max(alturaIzq, alturaDer);
}

// Libera la memoria del árbol recursivamente
void liberarArbol(Nodo* nodo) {
    if (nodo == nullptr) return;
    liberarArbol(nodo->izquierdo);
    liberarArbol(nodo->derecho);
    delete nodo;
}

int main() {
    
    Nodo* raiz = new Nodo(1);
    raiz->derecho = new Nodo(2);
    raiz->derecho->izquierdo = new Nodo(3);

    cout << "--- Prueba Ejercicio 3 ---" << endl;
    cout << "Altura esperada:  3" << endl;
    cout << "Altura calculada: " << calcularAltura(raiz) << endl;
    cout << "Altura de arbol nulo (esperado 0): " << calcularAltura(nullptr) << endl;

    liberarArbol(raiz);
    return 0;
}

File: Ejercicio5_Transformacion.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class NodoN {
    public int valor;
    public List<NodoN> hijos;

    public NodoN(int valor) {
        this.valor = valor;
        this.hijos = new ArrayList<>();
    }

    public NodoN(int valor, List<NodoN> hijos) {
        this.valor = valor;
        this.hijos = hijos;
    }
}

public class Ejercicio1_Basico {

    // Cuenta todos los nodos de un árbol N-ario de forma recursiva
 
    public static int contarNodos(NodoN raiz) {
        if (raiz == null)  // Caso base: nodo vacío no cuenta
            return 0;

        int total = 1;  // Cuenta el nodo actual

        // Recorre cada hijo y acumula recursivamente su cantidad de nodos
        for (NodoN hijo : raiz.hijos) {
            total += contarNodos(hijo);
        }

        return total;
    }

    public static void main(String[] args) {
        
        NodoN n5 = new NodoN(5);
        NodoN n6 = new NodoN(6);
        NodoN n2 = new NodoN(2, Arrays.asList(n5, n6));
        NodoN n3 = new NodoN(3);
        NodoN n4 = new NodoN(4);
        NodoN raiz = new NodoN(1, Arrays.asList(n2, n3, n4));

        System.out.println("--- Prueba Ejercicio 1 ---");
        System.out.println("Nodos esperados:  6");
        System.out.println("Nodos calculados: " + contarNodos(raiz));
    }
}
